File: framework/core/element_finder.py
"""
Element Finder - Handles element location and interaction with smart wait strategies
"""

from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException
import logging
import time

logger = logging.getLogger(__name__)


class ElementFinder:
    """Handles element finding with various strategies"""
    
    def __init__(self, driver, wait_timeout=20):
        self.driver = driver
        self.wait_timeout = wait_timeout
        self.wait = WebDriverWait(driver, wait_timeout)
        
    def find_element(self, locator_type, locator_value, timeout=None):
        """Find element with specified locator strategy"""
        timeout = timeout or self.wait_timeout
        wait = WebDriverWait(self.driver, timeout)
        
        try:
            by = self._get_by_type(locator_type)
            
            # First check if element exists
            element = wait.until(
                EC.presence_of_element_located((by, locator_value))
            )
            
            # Try to scroll to element if needed
            if not self._is_element_visible(element):
                self._scroll_to_element(element)
            
            # Wait for element to be clickable
            element = wait.until(
                EC.element_to_be_clickable((by, locator_value))
            )
            
            logger.info(f"Found element: {locator_type}={locator_value}")
            return element
            
        except TimeoutException:
            logger.error(f"Element not found: {locator_type}={locator_value}")
            # Try scrolling and searching
            if self._scroll_and_find(locator_type, locator_value):
                return self.find_element(locator_type, locator_value, timeout=5)
            raise
        except Exception as e:
            logger.error(f"Error finding element: {str(e)}")
            raise
    
    def find_elements(self, locator_type, locator_value, timeout=None):
        """Find multiple elements with specified locator strategy"""
        timeout = timeout or self.wait_timeout
        wait = WebDriverWait(self.driver, timeout)
        
        try:
            by = self._get_by_type(locator_type)
            elements = wait.until(
                EC.presence_of_all_elements_located((by, locator_value))
            )
            logger.info(f"Found {len(elements)} elements: {locator_type}={locator_value}")
            return elements
        except TimeoutException:
            logger.warning(f"No elements found: {locator_type}={locator_value}")
            return []
        except Exception as e:
            logger.error(f"Error finding elements: {str(e)}")
            return []
    
    def wait_for_element(self, locator_type, locator_value, timeout=None, condition='visible'):
        """Wait for element with specific condition"""
        timeout = timeout or self.wait_timeout
        wait = WebDriverWait(self.driver, timeout)
        by = self._get_by_type(locator_type)
        
        try:
            if condition == 'visible':
                element = wait.until(
                    EC.visibility_of_element_located((by, locator_value))
                )
            elif condition == 'clickable':
                element = wait.until(
                    EC.element_to_be_clickable((by, locator_value))
                )
            elif condition == 'present':
                element = wait.until(
                    EC.presence_of_element_located((by, locator_value))
                )
            else:
                raise ValueError(f"Unknown condition: {condition}")
            
            logger.info(f"Element is {condition}: {locator_type}={locator_value}")
            return element
        except TimeoutException:
            logger.error(f"Timeout waiting for element to be {condition}: {locator_type}={locator_value}")
            return None
        except Exception as e:
            logger.error(f"Error waiting for element: {str(e)}")
            return None
    
    def is_element_present(self, locator_type, locator_value, timeout=3):
        """Check if element is present without throwing exception"""
        try:
            self.wait_for_element(locator_type, locator_value, timeout=timeout, condition='present')
            return True
        except:
            return False
    
    def _get_by_type(self, locator_type):
        """Convert string locator type to By object"""
        locator_map = {
            'id': AppiumBy.ID,
            'accessibility_id': AppiumBy.ACCESSIBILITY_ID,
            'xpath': AppiumBy.XPATH,
            'class': AppiumBy.CLASS_NAME,
            'name': AppiumBy.NAME,
            'android_uiautomator': AppiumBy.ANDROID_UIAUTOMATOR,
            'ios_predicate': AppiumBy.IOS_PREDICATE_STRING,
            'ios_class_chain': AppiumBy.IOS_CLASS_CHAIN,
            'image': AppiumBy.IMAGE
        }
        
        locator_type = locator_type.lower()
        if locator_type not in locator_map:
            raise ValueError(f"Unknown locator type: {locator_type}")
        
        return locator_map[locator_type]
    
    def _is_element_visible(self, element):
        """Check if element is visible in viewport"""
        try:
            return element.is_displayed()
        except:
            return False
    
    def _scroll_to_element(self, element):
        """Scroll to make element visible"""
        try:
            if self.driver.capabilities.get('platformName', '').lower() == 'android':
                self.driver.execute_script('mobile: scrollToElement', {
                    'element': element.id,
                    'strategy': 'accessibility id',
                    'maxSwipes': 10
                })
            else:
                # iOS scroll
                self.driver.execute_script('mobile: scroll', {
                    'element': element.id,
                    'toVisible': True
                })
            logger.info("Scrolled to element")
            return True
        except Exception as e:
            logger.warning(f"Could not scroll to element: {str(e)}")
            return False
    
    def _scroll_and_find(self, locator_type, locator_value, max_scrolls=5):
        """Scroll and search for element"""
        platform = self.driver.capabilities.get('platformName', '').lower()
        
        for i in range(max_scrolls):
            if self.is_element_present(locator_type, locator_value, timeout=1):
                return True
            
            # Perform scroll
            if platform == 'android':
                self._scroll_down_android()
            else:
                self._scroll_down_ios()
            
            time.sleep(0.5)
        
        return False
    
    def _scroll_down_android(self):
        """Scroll down on Android"""
        try:
            size = self.driver.get_window_size()
            start_x = size['width'] // 2
            start_y = size['height'] * 0.8
            end_y = size['height'] * 0.2
            
            self.driver.swipe(start_x, start_y, start_x, end_y, duration=800)
        except Exception as e:
            logger.warning(f"Error scrolling: {str(e)}")
    
    def _scroll_down_ios(self):
        """Scroll down on iOS"""
        try:
            self.driver.execute_script('mobile: scroll', {'direction': 'down'})
        except Exception as e:
            logger.warning(f"Error scrolling: {str(e)}")
    
    def scroll_up(self):
        """Scroll up"""
        platform = self.driver.capabilities.get('platformName', '').lower()
        
        if platform == 'android':
            size = self.driver.get_window_size()
            start_x = size['width'] // 2
            start_y = size['height'] * 0.2
            end_y = size['height'] * 0.8
            self.driver.swipe(start_x, start_y, start_x, end_y, duration=800)
        else:
            self.driver.execute_script('mobile: scroll', {'direction': 'up'})

File: testzen-core/src/main/kotlin/com/testzen/core/execution/ElementFinder.kt
package com.testzen.core.execution

import com.testzen.core.config.TestZenConfig
import com.testzen.core.locator.*
import com.testzen.core.locator.smart.*
import com.testzen.core.model.Platform
import io.appium.java_client.AppiumBy
import io.appium.java_client.AppiumDriver
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.WebDriverWait
import org.slf4j.LoggerFactory
import java.time.Duration

/**
 * Handles element finding with optional self-healing and smart finding support.
 *
 * Responsible for:
 * - Finding elements using various locator strategies
 * - Coordinating with SelfHealingLocator when enabled
 * - Delegating to SmartElementFinder for intelligent element finding
 * - Building platform-specific locators
 * - Scroll-to-find functionality
 *
 * Single Responsibility: Element location and discovery orchestration.
 */
class ElementFinder(
    private val driver: WebDriver,
    private val config: TestZenConfig,
    private val platform: Platform,
    private val selfHealingLocator: SelfHealingLocator? = null,
    private val gestureHandler: GestureHandler? = null
) {
    private val logger = LoggerFactory.getLogger(ElementFinder::class.java)
    private val locatorGenerator = LocatorGenerator.forPlatform(platform)

    // Smart element finder for intelligent element finding
    private val smartFinder: SmartElementFinder by lazy {
        SmartElementFinder(
            driver = driver,
            platform = platform,
            config = SmartFinderConfig(
                minimumScore = config.smartFindMinimumScore,
                enableFallback = true,
                enableSpatialSearch = config.smartFindEnabled,
                enableCompoundResolution = config.smartFindEnabled
            )
        )
    }

    /**
     * Find an element by target name.
     *
     * @param target The element identifier (text, label, or accessibility ID)
     * @param timeoutMs Maximum time to wait for element
     * @param forTextEntry If true, include additional text entry field strategies
     * @return The found WebElement, or null if not found
     */
    fun find(target: String, timeoutMs: Long, forTextEntry: Boolean = false): WebElement? {
        // Use smart finder if enabled
        if (config.smartFindEnabled) {
            val action = if (forTextEntry) ActionType.ENTER_TEXT else ActionType.CLICK
            val result = findSmart(target, action)
            if (result != null) return result
        }

        // Use self-healing locator if enabled
        if (config.selfHealingEnabled && selfHealingLocator != null) {
            return findWithSelfHealing(target, timeoutMs, forTextEntry)
        }

        // Standard locator strategy
        return findStandard(target, timeoutMs, forTextEntry)
    }

    /**
     * Find an element for a specific action type using smart finding.
     *
     * @param target The element identifier
     * @param action The intended action type
     * @return The found WebElement, or null if not found
     */
    fun findForAction(target: String, action: ActionType): WebElement? {
        if (config.smartFindEnabled) {
            return findSmart(target, action)
        }

        // Fall back to standard find
        val forTextEntry = action == ActionType.ENTER_TEXT
        return findStandard(target, config.elementTimeoutMs, forTextEntry)
    }

    /**
     * Find an element using smart finding with spatial analysis.
     *
     * @param target The element identifier
     * @param action The intended action type
     * @return The found WebElement, or null if not found
     */
    fun findSmart(target: String, action: ActionType): WebElement? {
        return try {
            val result = smartFinder.findElement(target, action)
            when (result) {
                is FindResult.SingleElement -> {
                    logger.debug("Smart find success: target='$target', method=${result.candidate.findMethod}")
                    result.element
                }
                is FindResult.CompoundActions -> {
                    // For compound actions, return the first element
                    logger.debug("Smart find compound: ${result.actions.size} actions")
                    result.actions.firstOrNull()?.element?.element
                }
                is FindResult.NotFound -> {
                    logger.debug("Smart find not found: $target")
                    null
                }
            }
        } catch (e: Exception) {
            logger.debug("Smart find error: ${e.message}")
            null
        }
    }

    /**
     * Find element with compound action resolution.
     * Returns a FindResult which may contain multiple actions for compound UI.
     *
     * @param target The element identifier or instruction
     * @param action The intended action type
     * @return FindResult with element(s) or actions
     */
    fun findWithCompoundSupport(target: String, action: ActionType): FindResult {
        return try {
            smartFinder.findElement(target, action)
        } catch (e: Exception) {
            logger.debug("Compound find error: ${e.message}")
            FindResult.notFound(target, action, e.message ?: "Unknown error")
        }
    }

    /**
     * Find input field by label using spatial analysis.
     *
     * @param label The label text near the input field
     * @return The found input WebElement, or null if not found
     */
    fun findInputByLabel(label: String): WebElement? {
        if (config.smartFindEnabled) {
            val result = smartFinder.findInputField(label)
            if (result.success && result is FindResult.SingleElement) {
                return result.element
            }
        }
        return findStandard(label, config.elementTimeoutMs, forTextEntry = true)
    }

    /**
     * Find clickable element, traversing to parent if needed.
     *
     * @param target The element identifier
     * @return The found clickable WebElement, or null if not found
     */
    fun findClickable(target: String): WebElement? {
        if (config.smartFindEnabled) {
            val result = smartFinder.findClickable(target)
            if (result.success && result is FindResult.SingleElement) {
                return result.element
            }
        }
        return findStandard(target, config.elementTimeoutMs, forTextEntry = false)
    }

    /**
     * Find dropdown option by text.
     *
     * @param optionText The option text to find
     * @param dropdownElement Optional dropdown element to search within
     * @return The found option WebElement, or null if not found
     */
    fun findDropdownOption(optionText: String, dropdownElement: WebElement? = null): WebElement? {
        if (config.smartFindEnabled) {
            val result = smartFinder.findDropdownOption(optionText, dropdownElement)
            if (result.success && result is FindResult.SingleElement) {
                return result.element
            }
        }
        return findStandard(optionText, config.elementTimeoutMs, forTextEntry = false)
    }

    /**
     * Find element using self-healing locator with fallback strategies.
     */
    private fun findWithSelfHealing(target: String, timeoutMs: Long, forTextEntry: Boolean): WebElement? {
        val healer = selfHealingLocator ?: return findStandard(target, timeoutMs, forTextEntry)

        // Build additional locator hints for text entry fields
        val hints = if (forTextEntry) {
            locatorGenerator.generateForTextEntry(target)
        } else {
            emptyList()
        }

        return try {
            healer.findElement(target, hints)
        } catch (e: org.openqa.selenium.NoSuchElementException) {
            logger.debug("Self-healing locator failed for '$target', trying scroll strategy")

            // Try scrolling to find element
            if (config.scrollToFindElement && gestureHandler != null) {
                for (i in 1..config.maxScrollAttempts) {
                    gestureHandler.scroll(ScrollDirection.DOWN)
                    Thread.sleep(500)

                    try {
                        return healer.findElement(target, hints)
                    } catch (e2: org.openqa.selenium.NoSuchElementException) {
                        // Continue scrolling
                    }
                }
            }

            null
        }
    }

    /**
     * Standard element finding without self-healing.
     */
    private fun findStandard(target: String, timeoutMs: Long, forTextEntry: Boolean = false): WebElement? {
        val wait = WebDriverWait(driver, Duration.ofMillis(timeoutMs))
        val locators = buildSeleniumLocators(target, forTextEntry)

        for (locator in locators) {
            try {
                val element = wait.until(ExpectedConditions.presenceOfElementLocated(locator))
                if (element.isDisplayed) {
                    logger.debug("Found element with locator: $locator")
                    return element
                }
            } catch (e: Exception) {
                // Try next locator
            }
        }

        // Try scrolling to find element
        if (config.scrollToFindElement && gestureHandler != null) {
            for (i in 1..config.maxScrollAttempts) {
                gestureHandler.scroll(ScrollDirection.DOWN)
                Thread.sleep(500)

                for (locator in locators) {
                    try {
                        val element = driver.findElement(locator)
                        if (element.isDisplayed) {
                            return element
                        }
                    } catch (e: Exception) {
                        // Continue
                    }
                }
            }
        }

        return null
    }

    /**
     * Build Selenium By locators for standard element finding.
     * Note: For self-healing, use LocatorGenerator instead.
     */
    fun buildSeleniumLocators(target: String, forTextEntry: Boolean = false): List<By> {
        val locators = mutableListOf<By>()

        // Accessibility ID (highest priority)
        locators.add(AppiumBy.accessibilityId(target))

        // ID/Resource ID
        locators.add(By.id(target))

        // Platform-specific text locators
        if (driver is AppiumDriver) {
            // Android UIAutomator
            locators.add(AppiumBy.androidUIAutomator("new UiSelector().text(\"$target\")"))
            locators.add(AppiumBy.androidUIAutomator("new UiSelector().textContains(\"$target\")"))
            locators.add(AppiumBy.androidUIAutomator("new UiSelector().description(\"$target\")"))

            // iOS NSPredicate
            locators.add(AppiumBy.iOSNsPredicateString("label == '$target'"))
            locators.add(AppiumBy.iOSNsPredicateString("label CONTAINS '$target'"))
            locators.add(AppiumBy.iOSNsPredicateString("name == '$target'"))
        }

        // XPath with text
        locators.add(By.xpath("//*[@text='$target']"))
        locators.add(By.xpath("//*[contains(@text, '$target')]"))
        locators.add(By.xpath("//*[@content-desc='$target']"))
        locators.add(By.xpath("//*[contains(@content-desc, '$target')]"))
        locators.add(By.xpath("//*[contains(text(), '$target')]"))

        // For text entry, also look for editable fields near the label
        if (forTextEntry) {
            locators.add(By.xpath("//*[@text='$target']/following-sibling::*[1][@focusable='true']"))
            locators.add(By.xpath("//*[contains(@text, '$target')]/..//*[@focusable='true']"))
        }

        // CSS selectors for web
        locators.add(By.cssSelector("[placeholder='$target']"))
        locators.add(By.cssSelector("[aria-label='$target']"))
        locators.add(By.cssSelector("[name='$target']"))

        return locators
    }

    /**
     * Wait for element to appear or disappear.
     *
     * @param target The element identifier
     * @param timeoutMs Maximum wait time
     * @param shouldDisappear If true, wait for element to disappear
     * @return True if condition met, false otherwise
     */
    fun waitForElement(target: String, timeoutMs: Long, shouldDisappear: Boolean): Boolean {
        val wait = WebDriverWait(driver, Duration.ofMillis(timeoutMs))

        return try {
            if (shouldDisappear) {
                val locators = buildSeleniumLocators(target, false)
                for (locator in locators) {
                    try {
                        wait.until(ExpectedConditions.invisibilityOfElementLocated(locator))
                        return true
                    } catch (e: Exception) {
                        // Try next locator
                    }
                }
                true // Element not found = disappeared
            } else {
                val element = find(target, timeoutMs)
                element != null
            }
        } catch (e: Exception) {
            !shouldDisappear // Return false for appear, true for disappear
        }
    }
}
